src/libraries/System.Private.CoreLib/src/System/Diagnostics/UnreachableException.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Diagnostics
+{
+    /// <summary>
+    /// Exception thrown when the program executes an instruction that was thought to be unreachable.
+    /// </summary>
+    public sealed class UnreachableException : Exception
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="System.Diagnostics.UnreachableException"/> class with the default error message.
+        /// </summary>
+        public UnreachableException()
+            : base(SR.Arg_UnreachableException)
+        {
```

Comment:
This is only for truly unreachable cases, right? Indeed we're not verifying in our unit tests that anything throws it. And no future tests should ever expect it. In that case, why not add a Debug.Fail here. It would only affect our unit tests, as we don't ship debug bits. And the tests for this type itself would need to expect it, when run on debug builds. But it would help ensure that no future test is written that expects an UnreachableException, since that would indicate it is reachable.

Reply:
The `Debug.Fail` in the product code just before throwing UnreachableException could be removed then.

Reply:
> In that case, why not add a Debug.Fail here. It would only affect our unit tests, as we don't ship debug bits

There's probably relatively little downside to adding it, other than an inability to test the exception itself.  But there's also likely little upside, too: it would require running with a debug System.Private.CoreLib, which is rarely used to run library tests.

Reply:
> The Debug.Fail in the product code just before throwing UnreachableException could be removed then.

Only for the ones _in_ corelib.

Reply:
Ah! I overlooked it was in corelib. And I was not in a universe where https://github.com/dotnet/runtime/issues/13656 is fixed.